jwst/stpipe/tests/test_prepare_output.py

```diff
@@ -0,0 +1,240 @@
+)
+
+
+@pytest.fixture()
+def mock_copy(monkeypatch):
+    """Monkeypatch the ImageModel copy to raise an error if called."""
+
+    def _mock(*args, **kwargs):
+        raise RuntimeError("Copy was called")
+
+    monkeypatch.setattr(datamodels.image.ImageModel, "copy", _mock)
+
+
+@pytest.mark.parametrize("step", [PrepareOutputStep, PrepareOutputForceCopyStep])
+def test_prepare_output_from_str(step):
+    input_data = get_pkg_data_filename("data/science.fits", package="jwst.stpipe.tests")
+    result = step.call(input_data)
+
+    # Input is str, output is model
+    assert result is not input_data
+    assert isinstance(input_data, str)
+    assert isinstance(result, datamodels.ImageModel)
+    assert result.meta.cal_step.prepare_output == "COMPLETE"
+
+
+def test_prepare_output_from_str_no_open():
+    input_data = get_pkg_data_filename("data/science.fits", package="jwst.stpipe.tests")
+
+    # Step does not call datamodels.open for the input string
+    result = PrepareOutputNoOpenStep.call(input_data)
+
+    assert result is input_data
+    assert isinstance(input_data, str)
+    assert isinstance(result, str)
+
+
+def test_prepare_output_from_str_force_copy_no_open(monkeypatch):
+    input_data = get_pkg_data_filename("data/science.fits", package="jwst.stpipe.tests")
+
+    # Raises a type error: input is not opened so it can't be copied
+    with pytest.raises(TypeError, match="Copy is not possible"):
+        PrepareOutputForceCopyNoOpenStep.call(input_data)
+
+
+@pytest.mark.parametrize("step", [PrepareOutputStep, PrepareOutputForceCopyStep])
+def test_prepare_output_from_model(step):
+    input_data = datamodels.ImageModel()
+    result = step.call(input_data)
+
+    # Output is a copy of the input
+    assert result is not input_data
+    assert isinstance(result, datamodels.ImageModel)
+    assert result.meta.cal_step.prepare_output == "COMPLETE"
+    assert not hasattr(input_data.meta.cal_step, "prepare_output")
+
+
+def test_prepare_output_from_model_no_copy():
+    input_data = datamodels.ImageModel()
+    result = PrepareOutputNoCopyStep.call(input_data)
+
+    # Output is the same as the input
+    assert result is input_data
+    assert isinstance(result, datamodels.ImageModel)
+
+    assert result.meta.cal_step.prepare_output == "COMPLETE"
+    assert input_data.meta.cal_step.prepare_output == "COMPLETE"
+
+
+@pytest.mark.parametrize("step", [PrepareOutputStep, PrepareOutputForceCopyStep])
+def test_prepare_output_from_container(step):
+    container = ModelContainer([datamodels.ImageModel()])
+    input_model = container[0]
+    result = step.call(container)
+
+    # Output is a copy of the input
+    assert result is not container
+    assert result[0] is not input_model
+    assert isinstance(result, ModelContainer)
+    assert isinstance(result[0], datamodels.ImageModel)
+    assert result[0].meta.cal_step.prepare_output == "COMPLETE"
+    assert not hasattr(input_model.meta.cal_step, "prepare_output")
```

Comment:
here and for other instances below, it probably doesn't matter in this case but would it be more technically correct to use model.hasattr() to get around the hasattr bug?

Reply:
yea I figured that would be the behavior.  Thanks for changing it anyway.  I was thinking more of the case that someone might find this code, copy-paste it, change the attribute to a schema-defined one, and then get confused about why their code doesn't work